onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorShape.cpp

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "precomp.h"
+
+namespace Dml
+{
+
+class DmlOperatorShape : public DmlOperator, ShapeHelper
+{
+public:
+    DmlOperatorShape(const MLOperatorKernelCreationContext& kernelCreationContext)
+    :   DmlOperator(kernelCreationContext),
+        ShapeHelper(kernelCreationContext, kernelCreationContext.GetTensorShapeDescription())
+    {
+        ML_CHECK_VALID_ARGUMENT(kernelCreationContext.GetInputCount() == 1, "Shape expects 1 input tensor.");
+        ML_CHECK_VALID_ARGUMENT(kernelCreationContext.GetOutputCount() == 1, "Shape expects 1 output tensor.");
+
+        m_startIndex = kernelCreationContext.GetOptionalAttribute<int32_t>(AttrName::Start, 0);
+
+        if (m_startIndex != 0)
+        {
+            // "start" is provided and is non-default (default is 0)
+            m_needsSlicing = true;
+        }
+
+        if (kernelCreationContext.HasAttribute(AttrName::End, MLOperatorAttributeType::Int))
+        {
+            m_endIndex = kernelCreationContext.GetAttribute<int64_t>(AttrName::End);
+            m_needsSlicing = true;
+        }
+
+        DmlOperator::Initialize(kernelCreationContext);
+    }
+
+    // Takes a tensor as input and outputs a 1D int64 tensor containing the shape of the input tensor.
+    void Compute(const MLOperatorKernelContext& kernelContext)
+    {
+        std::vector<IMLOperatorTensor*> inputTensors = GetInputTensors(kernelContext);
+        std::vector<IMLOperatorTensor*> outputTensors = GetOutputTensors(kernelContext);
+        assert(inputTensors.size() == 1);
+        assert(outputTensors.size() == 1);
+
+        const IMLOperatorTensor* inputTensor = inputTensors[0];
+        IMLOperatorTensor* outputTensor = outputTensors[0];
+        assert(outputTensor->GetDimensionCount() == 1);
+
+        uint32_t inputDimCount = inputTensor->GetDimensionCount();
+        uint32_t outputSize;
+        THROW_IF_FAILED(outputTensor->GetShape(1, &outputSize));
+        assert(outputSize == inputDimCount);
+
```

Comment:
Would it be easier to just use MLOperatorTensor?

```
MLOperatorTensor outputTensor(outputTensors.front());
const std::vector<uint32_t> outputShape = outputTensor.GetShape();

if (m_needsSlicing)
{
    int64_t start = HandleNegativeAxis(m_startIndex, outputDimCount);
    int64_t end = HandleNegativeAxis(m_endIndex, outputDimCount);
    outputShape.erase(outputShape.begin() + end);
    outputShape.erase(outputShape.begin(), outputShape.begin() + start);
}

```

Reply:
And per the other comment, the start/end resolution should be in inference code AFAICS.

Reply:
It doesn't do the same thing. `HandleNegativeAxis` returns a `uint32_t`, but according to the CPU implementation, it's possible that `abs(startIndex) > inputDimCount` when `startIndex` is negative (although I don't know how likely it is). So it needs to clamp the value instead of blindly converting it to `uint32_t`.